

# Recipe #4:  Taste Profiles



For taste profiles, we need to index the following data-shape:  

(**user**, **query[]**, **document[]**)

In [1]:
import sys
!conda install --yes --prefix {sys.prefix} elasticsearch jsonlines jsonpath-ng python-slugify

Solving environment: done

# All requested packages already installed.



In [1]:
import pandas
import json
from extract import extract, parse_json
from taste.transform import reshape, preload
from index import create_index, populate_index, query_index, index_mapping, query_template, peek, get


## Source data 


In [2]:
source = extract('./data/real_estate_queries.tsv')

# peek
source.head(5)

user                                              query  \
0  1337           national real estate settlement services   
1  1337           national real estate settlement services   
2  1337  sunbury pennsylvania real estate settlement se...   
3  1337                             integrated real estate   
4  1337                             integrated real estate   

             timestamp  position                       document         state  
0  2006-03-14 15:59:13       1.0         http://www.realtms.com           NaN  
1  2006-03-14 15:59:13       7.0                http://dmoz.org           NaN  
2  2006-03-14 16:06:28      14.0       http://pa.optimuslaw.com  pennsylvania  
3  2006-03-27 14:52:29       1.0  http://www.integratedreal.com           NaN  
4  2006-03-27 14:52:29       2.0         http://www.irisnet.net           NaN


##  Transform to fit the recipe


In [3]:
reshaped = reshape(source)

# peek
reshaped.head(50)

user                                              query  \
0     475  [real estates lookup in charlotte, real estate...   
1     756  [chesapeake real estate assesser, virginia bea...   
2    1133   [huntsville real estate, huntsville real estate]   
3    1337  [national real estate settlement services, nat...   
4    1338  [real estate chesapeake appreciation 2005, rea...   
5    2470                       [harbor country real estate]   
6    2761  [norco real estate, norco real estate, corona ...   
7    3769                          [little rock real estate]   
8    4781             [tucker group real estate and florida]   
9    4945  [real estate courses, real estate courses, fis...   
10   5070                               [powell real estate]   
11   5826                  [illinois commercial real estate]   
12   7764  [galveston beach opinions real estate, galvest...   
13   9780  [hamilton montana real estate, hamilton montan...   
14   9993  [nys real estate exam, www. real estate manage...   
15  10008  [cape cod real estate.com, new hampshire me re...   
16  10215  [real estate orlando florida, real estate orla...   
17  11663  [real estate findlay ohio, rooney real estate ...   
18  12653  [real estate for dummies, real estate for dumm...   
19  13097                        [real estate home warranty]   
20  13119  [what is a real estate agent suppose to do, wh...   
21  13379  [real estate northern liberties philadelphia, ...   
22  13577                                [mogen real estate]   
23  14295              [potomac real estate coldwell banker]   
24  15997      [luxury real estate, real estate port richey]   
25  16790  [denver co real estate cherry creek, denver co...   
26  18987                           [college of real estate]   
27  19058               [meridian ms commercial real estate]   
28  19166                         [ewa beach ha real estate]   
29  19913         [bashore real estate, bashore real estate]   
30  19975                  [blairsville georgia real estate]   
31  20104   [sea pines real estate, savannah ga real estate]   
32  20320  [real estate for sale saratoga new york, real ...   
33  21901                  [dijon france real estate agents]   
34  22304                   [carollo real estate of bayside]   
35  22516  [real estate orange tip irvine calif 92604, re...   
36  22698  [real estate pedley california, real estate pe...   
37  24122                          [texas real estate agent]   
38  24537                      [keller williams real estate]   
39  25003               [mosley-flint school of real estate]   
40  25804         [mcneary real estate, mcneary real estate]   
41  27371                      [hampton country real estate]   
42  27949  [texas instruments ba real real estate, texas ...   
43  28538   [prudential real estate, prudential real estate]   
44  28829           [keller real estate, keller real estate]   
45  29720  [real estate irs withholding calculator califo...   
46  30011  [real estate in rosemount mi, real estate in r...   
47  30054                         [howard hanna real estate]   
48  30996         [real estate classes near hillsborough nc]   
49  34363  [real estate in covington la, real estate in c...   

                                             document  
0   [http://meckcama.co.mecklenburg.nc.us, http://...  
1   [http://www.chesapeake.va.us, http://www.vbgov...  
2   [http://www.valleymls.com, http://www.valleyml...  
3   [http://www.realtms.com, http://dmoz.org, http...  
4   [http://realtytimes.com, http://realtytimes.co...  
5                      [http://www.harborcountry.org]  
6   [http://www.neighborhoodscout.com, http://www....  
7                          [http://cityguide.aol.com]  
8                        [http://www.bizjournals.com]  
9   [http://www.online-education.net, http://www.u...  
10                  [http://www.powellrealestate.com]  
11                          [http://www.cityfeet.com]  
12  [http://

E.g. User 756 searched for "chesapeake real estate assesser" and "virginia beach" and at some point via search clicked on http://www.chesapeake.va.us and http://www.vbgov...

In [4]:
load_ready = preload(reshaped)

# peek
[print(j) for j in load_ready[0:10]]


{'index': {'_index': 'taste', '_id': 475}}
{'user': 475, 'query': ['real estates lookup in charlotte', 'real estates lookup in lexington', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in winston-salem', 'real estates lookup in winston-salem', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in rural-hall', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup in charlotte', 'real estates lookup and phone numbers', 'real estates lookup and phone numbers', 'real estates lookup in charlotte', 'real estates lookup in winston-salem', 'real estates lookup in winston-salem', 'real estates lookup in charlotte', 'real 

[None, None, None, None, None, None, None, None, None, None]


## Load the index


In [5]:
create_index(index='taste', mapping_file='taste/index_definition.json')

index_mapping(index='taste')

Mapping for  taste
{
  "taste": {
    "mappings": {
      "properties": {
        "document": {
          "type": "keyword"
        },
        "query": {
          "type": "text",
          "fields": {
            "raw": {
              "type": "keyword"
            },
            "sayt": {
              "type": "search_as_you_type",
              "analyzer": "query_log_analyzer",
              "max_shingle_size": 3
            }
          },
          "analyzer": "query_log_analyzer"
        },
        "user": {
          "type": "keyword"
        }
      }
    }
  }
}


In [6]:
response = populate_index(body=load_ready)


## Taste profile for one user


In [16]:
profile = get('10008', 'taste')

print(json.dumps(profile, indent=2))

{
  "_index": "taste",
  "_type": "_doc",
  "_id": "10008",
  "_version": 1,
  "_seq_no": 15,
  "_primary_term": 1,
  "found": true,
  "_source": {
    "user": 10008,
    "query": [
      "cape cod real estate.com",
      "new hampshire me real estate",
      "new hampshire me real estate",
      "new hampshire me real estate",
      "nancy clayton real estate falmouth ma",
      "nancy clayton real estate falmouth ma",
      "nancy clayton real estate falmouth ma",
      "nancy clayton real estate falmouth ma",
      "new hampshire maine real estate",
      "new hampshire maine real estate",
      "new hampshire maine real estate",
      "cape cod real estate associates",
      "marcy broden era key real estate framingham ma",
      "hudson nh real estate",
      "salem nh real estate",
      "salem nh real estate"
    ],
    "document": [
      "http://www.capecodrealestate.com",
      "http://www.seacoastrealestate.com",
      "http://www.nneren.com",
      "http://www.nneren.com",


## What other users share these interests? 

In [8]:
peek(filename='taste/query_template.mustache')

{
  "size": 0,
  "query": {
    "more_like_this" : {
      "fields" : ["query", "document"],
      "like" : [
      {
        "_index" : "taste",
        "_id" : "{{user}}"
      }
      ],
      "min_term_freq" : 1, 
      "max_query_terms" : 8
    }
  },
  "aggs": {
    "query_recommendations": {
      "terms": {
        "field": "query.raw",
        "size": {{^size}}50{{/size}}{{size}}
      }
    },
    "document_recommendations": {
      "terms": {
        "field": "document",
        "size": {{^size}}50{{/size}}{{size}}
      }
    }
  }
}



##  And what queries do **they** use?

In [36]:
# Run the query
result = query_template(index='taste', 
               template_file='taste/query_template.mustache',
               params={ 'user': '10008', 'size': 50 }
              )

# Parse the candidate queries out of the response
query_suggestion_candidates = parse_json(result, jsonpath="aggregations.query_recommendations.buckets[*].key")

# Remove the candidate queries already run by the user
query_suggestion = [s for s in query_suggestion_candidates if s not in profile['_source']['query']]

# Show
pandas.DataFrame({'Suggested Queries': query_suggestion}).head(25)

Suggested Queries
0                           new hampshire real estate
1                                   maine real estate
2                              bath maine real estate
3                         brunswick maine real estate
4                          north carolina real estate
5                                  oregon real estate
6                              prudential real estate
7                          south carolina real estate
8                          united country real estate
9                                   98903 real estate
10                     affordable real estate florida
11                          agua dulce ca real estate
12                             ahoskie nc real estate
13               alberta banff commercial real estate
14                               alliance real estate
15                         almaden valley real estate
16                            amesbury ma real estate
17                       antique colonial real estate
18           antique colonial real estate sales maine
19                  antique real estate new hampshire
20                               aol real estate sale
21                             arco idaho real estate
22                       arizona real estate for sale
23                arizona real estate new communities
24  arizona real estate new communities under 200 000

In [37]:
# Parse the candidate documents out of the response
document_suggestion_candidates = parse_json(result, jsonpath="aggregations.document_recommendations.buckets[*].key")

# Remove the candidate documents already visited by the user
document_suggestion = [s for s in document_suggestion_candidates if s not in profile['_source']['document']]

# Show 
pandas.DataFrame({'Suggested Documents': document_suggestion}).head(25)

Suggested Documents
0                 http://www.realtor.com
1              http://www.realestate.com
2                http://www.homegain.com
3            http://realestate.yahoo.com
4               http://www.househunt.com
5               http://www.city-data.com
6             http://www.harmonhomes.com
7                http://homes.point2.com
8            http://www.homesandland.com
9         http://www.hometownlocator.com
10          http://www.unitedcountry.com
11                      http://reals.com
12              http://www.beangroup.com
13                http://www.bobvila.com
14                http://www.globere.com
15         http://www.gmacrealestate.com
16                http://www.loopnet.com
17               http://www.newhomes.com
18               http://www.nhliving.com
19         http://real-estate-agents.com
20                   http://valuecom.com
21                 http://www.4homes.com
22              http://www.century21.com
23  http://www.epoweredprofessionals.com
24           http://www.escapeartist.com